### DISCUSSION TITLE
How to type annotate never ending generators?

### DISCUSSION THREAD

**Discussion by kumaraditya303** [python/typing#960] [Q&A]
How to type annotate never ending generators?

My current code:

```py

from typing import Generator, NoReturn


def func() -> Generator[int, None, NoReturn]:
    c = 0
    while True:
        c += 1
        yield c


def main() -> NoReturn:
    for i in func():
        print(i)
    # <--- This will never execute 

main()


```

**Comment by srittau** [comment]
I don't think this is currently possible, or at least well-defined, although your solution feels "natural" to me. Maybe this is something that type checkers could implement?

**Comment by kumaraditya303** [comment]
See https://github.com/microsoft/pyright/issues/2598 cc @erictraut

**Comment by kumaraditya303** [comment]
There is  no example of NoReturn with generators in PEP 484 either https://www.python.org/dev/peps/pep-0484/#the-noreturn-type

**Comment by kumaraditya303** [comment]
Perhaps the authors of PEP 484 can clarify here what should NoReturn do when used in generators?

**Reply by srittau** [reply]
This is certainly a case that's not covered by PEP 484, but a new feature. Whether this needs a new PEP or not, I'm not sure. But it would at least require buy-in of type checker authors.

**Reply by JelleZijlstra** [reply]
Yes. I think the right way forward here is:
- Open an issue on this repo outlining the proposed change
- Get buyin from major type checker authors (mypy, pyright, pyre, pytype)
- Update https://docs.python.org/3.10/library/typing.html#typing.NoReturn to mention this use case
- Implement support in type checkers

(Not necessarily in this exact order)
